src/gui/wizard/firstlay.cpp

```diff
@@ -2,627 +2,191 @@
-            return 100;
-        }
+StateFncData StateFnc_FIRSTLAY_FILAMENT_UNLOAD(StateFncData last_run) {
+    switch (gui_dlg_unload_forced()) {
+    case DLG_OK:
+        return StateFncData(WizardState_t::FIRSTLAY_FILAMENT_LOAD, WizardTestState_t::PASSED);
+    case DLG_ABORTED:
+        return StateFncData(WizardState_t::FIRSTLAY_FILAMENT_ASK, WizardTestState_t::PASSED);
     }
+    return FIRSTLAY_FILAMENT_done();
+}
 
-    int remaining_lines;
-    switch (p_screen->state) {
-    case _FL_INIT:
-        p_screen->Z_offset = z_offset;
-        wizard_init_screen_firstlay(id_body, p_screen, p_data);
-    #if DEBUG_TERM == 1
-        term_printf(&p_screen->terminal, "INITIALIZED\n");
-        p_screen->term.id.Invalidate();
-    #endif
-        _set_gcode_first_lines();
-        p_screen->state = _FL_GCODE_HEAD;
-        marlin_error_clr(MARLIN_ERR_ProbingFailed);
-    #if DEBUG_TERM == 1
-        term_printf(&p_screen->terminal, "HEAD\n");
-        p_screen->term.id.Invalidate();
-    #endif
-        break;
-    case _FL_GCODE_HEAD:
-        //have to wait to next state after MBL to check error
-        if (line_head > G29_pos && marlin_error(MARLIN_ERR_ProbingFailed)) {
-            marlin_error_clr(MARLIN_ERR_ProbingFailed);
-            marlin_gcode("G0 Z30"); //Z 30mm
-            marlin_gcode("M84");    //Disable steppers
-            if (wizard_msgbox(_("Mesh bed leveling failed?"), MSGBOX_BTN_RETRYCANCEL, 0) == MSGBOX_RES_RETRY) {
-                //RETRY
-                line_head = G28_pos;
-            } else {
-                //CANCEL
-                p_data->state_print = _TEST_FAILED;
-                return 100;
-            }
-        }
-    #if DEBUG_TERM == 0
-        remaining_lines = _run_gcode_line(&line_head, head_gcode,
-            head_gcode_sz);
-    #else
-        remaining_lines = _run_gcode_line(&line_head, head_gcode,
-            head_gcode_sz, &p_screen->term);
-    #endif
-        if (remaining_lines < 1) {
-            p_screen->state = _FL_GCODE_BODY;
-    #if DEBUG_TERM == 1
-            term_printf(&p_screen->terminal, "BODY\n");
-            p_screen->term.Invalidate();
-    #endif
-            p_screen->Z_offset_request = 0; //ignore Z_offset_request variable changes until now
-            p_screen->spin_baby_step.color_text = COLOR_ORANGE;
-            p_screen->spin_baby_step.Invalidate();
+StateFncData StateFnc_FIRSTLAY_MSBX_CALIB(StateFncData last_run) {
+    static const char en_text[] = N_("Now, let's calibrate the distance between the tip of the nozzle and the print sheet.");
+    string_view_utf8 translatedText = _(en_text);
+    MsgBox(translatedText, Responses_Next);
+    return last_run.PassToNext();
+}
+
+StateFncData StateFnc_FIRSTLAY_MSBX_USEVAL(StateFncData last_run) {
+    //show dialog only when values are not equal
+    float diff = marlin_vars()->z_offset - z_offset_def;
+    if ((diff <= -z_offset_step) || (diff >= z_offset_step)) {
+        char buff[20 * 7];
```

Comment:
Can you please define what 20*7 means?

Reply:
will try, I am not sure, it is moved code

Reply:
Sorry, cannot. Not my code, will not change it, because it worked ...